pages/_app.js

```diff
@@ -1,25 +1,51 @@
-import NextApp from "next/app";
-import React from "react";
+import React, { useEffect, useState } from "react";
 import { ThemeProvider as MuiThemeProvider } from "@material-ui/core/styles";
 import CssBaseline from "@material-ui/core/CssBaseline";
 import { ContextProviders } from "context/ContextProviders";
 import { ComponentsProvider } from "@reactioncommerce/components-context";
 import components from "custom/componentsContext";
 import theme from "custom/reactionTheme";
+import { useRouter } from "next/router";
+import analytics from "lib/analytics";
+import { AnalyticsProvider } from "use-analytics";
+import PropTypes from "prop-types";
+import { decodeOpaqueId } from "lib/utils/decoding";
 
-export default class App extends NextApp {
-  componentDidMount() {
+const App = (props) => {
+  const router = useRouter();
+  const [previewUrl, setPreviewUrl] = useState(null);
+
+  useEffect(() => {
     // Remove the server-side injected CSS.
     const jssStyles = document.querySelector("#jss-server-side");
     if (jssStyles && jssStyles.parentNode) {
       jssStyles.parentNode.removeChild(jssStyles);
     }
-  }
 
-  render() {
-    const { Component, pageProps, ...rest } = this.props;
+    const handleRouteChange = (url) => {
+      if (previewUrl === url) return true;
+      setPreviewUrl(url);
+      const { isFallback, query } = router;
+      const { shop } = props.pageProps;
+      const shopDecoded = shop?._id && decodeOpaqueId(shop._id);
+
+      return analytics.page({
+        isFallback,
+        path: url,
+        query,
+        shopId: shopDecoded?.id
+      });
+    };
+    router.events.on("routeChangeComplete", handleRouteChange);
+    return () => {
+      router.events.off("routeChangeComplete", handleRouteChange);
+    };
+  }, [router.events]);
```

Comment:
My preference would be to avoid adding more complexity right to this file, instead creating a separate hook for this, which could also include the logic you added in `lib/analytics/index.js` to initialize Analtics and return that instance.

So in _app.js you would only have something like e.g.  `const analytics = useInitAnalytics()` and pass that instance to `AnalyticsProvider` as you do right now.
Inside of that hook, you could declare that effect as part of the lifecycle, but still have the initialization of the analytics module done global to the file.

Some other things regarding the the call to `analytics.page`:

- I believe this will be executed on the server side during page creation aswell and is also supportedby the library, but not what we want, so we should probably check for `typeof window !== "undefined"` before and only execute that conditionally (no need to return it as done above).

- I could observe some issues with the nextjs routing event. as that somehow fired before the changed url was propagated properly. I don't recall the exact details, but the problem was there for quite some time, and still occured for the latest version I tested with around 4 weeks ago.

So I ended up with something like:
```js
if (typeof window !== "undefined") {
        setTimeout(() => {
          analytics.page({
            path: url,
          });
        }, 1000);
}
```

- I guess `previewUrl` is set to avoid duplicates, but A. I find the name a bit misleading, B. I Wonder if that is necessary as I wouldn't expect the event to fire multiple times for no reason, but if there indeed is a route transition to the same route, I see no issue with also tracking that.

- I believe `isFallback`, `query` and `shopId` are all non-default values and wonder if they are used by the included providers. While it is fine in general to pass custom pagedata here, did you check if the Providers actually parse it? I recall at least the Google Analytics plugin to provide some logic to pass them as custom "dimensions", so maybe it works fine.
Either way, I would prefer to not decode the shop Id here and pass it as-is, in the end the base64-encoded one is the "real" client-facing ID we have.

Reply:
thanks a lot! 🙌